### PR TITLE
Easy sandbox deploys

### LINKED ISSUE #130: One-script environment deploys
## Story

As a ***engineer***,
I would like to ***have a single-script deployment***,
So that I can ***easily create review apps for my PRs***

## Solution

```[tasklist]
### Tasks
- [x] Single script deploy via terraform
```

## Acceptance Criteria

- [x] A new app can be deployed to cloud.gov using a single command (after some config writing, maybe)

### PR DESCRIPTION
**Pull request by rahearn**
# 🎫 Addresses issue: #130 

closes #130 

## 🛠 Summary of changes

* Create a `terraform.sh` script that can deploy a sandbox 
* Create a sandbox_bot module to automatically create a service account to run terraform.sh
* Updates to the terraform generator so that it can still work when being used in the `sandbox-gsa` organization (previously, everything would fail because you couldn't use `OrgManager` permissions) (this ended up being a bigger lift than I meant it to)

## 📜 Testing Plan

How would a peer test this work?

Similar to #128, create a new app and deploy with`./terraform.sh`

* checkout branch
* run: export RT_DEV=true
* run: exe/rails_template_18f new ../my-rails-app and answer the questions. use gsa-tts-devtools-prototyping as the cg org, and make up a new space name for prod and staging spaces. They should not yet exist.
* move over to the new app and follow the terraform readme to deploy the app to cloud.gov
* repeat as above, but use your `sandbox-gsa` org and space

Possible gotcha - in order to use the script to deploy to staging or production environments, you need to tell it to use a different secrets file than it expects. Do this with:

```bash
./terraform.sh -e staging -c apply -- -var-file=secrets.cicd.tfvars
```

### CODE REVIEW

**Review by zjrgov — COMMENTED** [review]
Code changes look good! I think I should have time to test it today.

File: template.rb [hunk 1]
```diff
@@ -109,6 +118,8 @@ def print_announcements
 run_db_setup = yes?("Run db setup steps? (y/n)")
 
 ## Start of app customizations
+insert_into_file "Gemfile", "\n#{gem_ruby_entry}\n", after: /^source "https.*\n/
```
Comment (rahearn) [inline template.rb:121]: @zjrgov this will fix the deploy issue of cloud.gov not knowing what version of ruby to use

File: lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt [hunk 2]
```diff
@@ -2,6 +2,21 @@
 
 This directory holds the terraform module for maintaining the system infrastructure and deploying the application.
 
+<% unless terraform_manage_spaces? %>
+## READ ME FIRST
+
+Due to users not having `OrgManager` permission in the `sandbox-gsa` organization, this version of the terraform module
+is very limited and some of the rest of this README does not apply.
+
+When you are ready to move the application to a non-sandbox cloud.gov organization, please re-run the terraform generator with
```
Comment (zjrgov) [inline lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt:11]:
```suggestion
When you are ready to move the application to a non-sandbox cloud.gov organization, please re-run the terraform generator with…
```

File: lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt [hunk 3]
```diff
@@ -2,6 +2,21 @@
...
+
+```bash
+bin/rails generate rails_template18f:terraform --cg-org=<ORG_NAME> --cg-staging=<STAGING_SPACE_NAME> --cg-prod=<PRODUCTION_SPACE_NAME>
+```
+
+to take full advantage of the generator, and then re-run your CI generator of choice to add production terraform plan and apply steps to your workflow.
```
Comment (zjrgov) [inline lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt:17]:
```suggestion
…to take full advantage of the generator, and then re-run your CI generator of choice to add production terraform plan and apply steps to your workflow.
```

File: lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt [hunk 4]
```diff
@@ -11,76 +26,47 @@ create credentials files so developers can use that s3 bucket to create their ow
 
 These steps only need to be run once per project.
 
-1. `cd bootstrap`
-1. Add any users who should have access to the terraform state bucket to `users.auto.tfvars`
-1. Run `cf login --sso` to log in to cloud.gov
+1. `cd bootstrap`<% if terraform_manage_spaces? %>
+1. Add any users who should have access to the terraform state bucket to `users.auto.tfvars`<% end %>
```
Comment (zjrgov) [inline lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt:30]: [question] is there any reason people who aren't just testing the template might want to use a basic-sandbox s3 bucket? Was wondering whether a mention of it here would make sense.

File: lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt [hunk 5]
```diff
@@ -2,6 +2,21 @@
 
 This directory holds the terraform module for maintaining the system infrastructure and deploying the application.
 
+<% unless terraform_manage_spaces? %>
+## READ ME FIRST
+
+Due to users not having `OrgManager` permission in the `sandbox-gsa` organization, this version of the terraform module
+is very limited and some of the rest of this README does not apply.
```
Comment (zjrgov) [inline lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt:9]: [suggestion] I think it could be good to either directly specify what parts to run or not, or to cover all necessary exclusions with more conditionals and reword this to more of an FYI for when people migrate out of sandbox.

File: lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt [hunk 6]
```diff
@@ -11,76 +26,47 @@ create credentials files so developers can use that s3 bucket to create their ow
...
+    ./terraform.sh -e sandbox-<NAME> -c apply
     ```
 
 1. Optional: tear down the sandbox if it does not need to be used anymore
     ```bash
-    terraform destroy -var-file=sandbox-name.tfvars -var rails_master_key=anything
-    # <SPACE_NAME> and <ACCOUNT_NAME> have the same values as used above.
-    ../../bin/ops/destroy_service_account.sh -s <SPACE_NAME> -u <ACCOUNT_NAME>
-    rm secrets.auto.tfvars
+    ./terraform.sh -e sandbox-<NAME> -c destroy
```
Comment (zjrgov) [inline lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt:69]: [question] for sandbox-gsa, should all `sandbox-<NAME>` instead be `staging`, and the note about copying `staging.tfvars` get excluded?

errr… I was thinking this when I was doing the deployment, but I guess I'm not actually sure how people would use this in a sandboxed context. Will there be multiple users trying to work on an app in one person's sandbox?

File: template.rb [hunk 7]
```diff
@@ -107,6 +118,8 @@ def print_announcements
...
+
 template "README.md", force: true
 register_announcement("Documentation", <<~EOM)
```
Comment (zjrgov) [inline template.rb:124]: [nitpick (nb)] there's a weird gap between this section of the output and the next when you are running for `sandbox-gsa`

Screenshot 2025-01-02 at 6 19 30 PM

**Review by zjrgov — DISMISSED** [review]
This all looks good, and I was able to run through for both the prototyping space, my sandbox, and from the sandbox back to prototyping without issues (on the part of the template).

I was reminded about something, which is that I think either the generated README.md or the templater's output should include some mention about going into the `./terraform` directory and looking at `./terraform/README.md`. I think I maybe brought this up before though? and that maybe you had a reason not to?

File: lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt (hunk 6 above)
Comment (rahearn) [reply lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt:69]: In general, people should not be collaborating in sandbox-gsa spaces, but it can and does happen, though for the most part I'm not interested in trying to run multiple copies of the app in a single sandbox space cause that'll get weird fast.

I think for now, at least, this part is covered by the "some parts of this file may not apply" disclaimer that's put in the top of the readme

File: lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt (hunk 4 above)
Comment (rahearn) [reply lib/generators/rails_template18f/terraform/templates/terraform/README.md.tt:30]: Nah - basic-sandbox is way more useful for people testing out the template than for actual users trying to build real infrastructure. The pain of deleting the bucket is far outweighed by the benefit of not accidentally deleting your state file.

**Comment by rahearn** [comment]
> I think I maybe brought this up before though? and that maybe you had a reason not to?

I think I just forgot to add it. Added this time along with a couple other doc changes based on your approval, so it needs a re-approval.

File: templates/README.md.tt [hunk 8]
```diff
@@ -110,13 +110,16 @@ To enable automatic ruby linting on every `git commit` follow the instructions a
 Each environment has dependencies on a PostgreSQL RDS instance managed by cloud.gov.
 See [cloud.gov docs](https://cloud.gov/docs/services/relational-database/) for information on RDS.
 
+Terraform is used to deploy the application and supporting services. See [terraform/README.md](./terraform/README.md)
+for more infromation on how to set up your terraform backend and deploy the app.
```
Comment (zjrgov) [inline templates/README.md.tt:114]: [typo]
```suggestion
for more information on how to set up your terraform backend and deploy the app.
```

**Review by zjrgov — APPROVED** [review]
